File: src/Niche/Prolog/factory_method.pl
create_postgres(postgres).
create_mysql(mysql).

connect(postgres) :- writeln('PostgreSQL connect').
query(postgres) :- writeln('PostgreSQL query').
connect(mysql) :- writeln('MySQL connect').
query(mysql) :- writeln('MySQL query').

use_database(CreatePredicate) :-
    Goal =.. [CreatePredicate, Database],
    call(Goal),
    connect(Database),
    query(Database).

run :-
    use_database(create_postgres),
    use_database(create_mysql).

File: src/Scripting/PHP/factory_method.php
<?php

interface FactoryMethodDatabase
{
    public function connect(): void;
    public function query(): void;
}

final class FactoryMethodPostgres implements FactoryMethodDatabase
{
    public function connect(): void { echo "PostgreSQL connect\n"; }
    public function query(): void { echo "PostgreSQL query\n"; }
}

final class FactoryMethodMySql implements FactoryMethodDatabase
{
    public function connect(): void { echo "MySQL connect\n"; }
    public function query(): void { echo "MySQL query\n"; }
}

abstract class FactoryMethodCreator
{
    abstract protected function createDatabase(): FactoryMethodDatabase;

    final public function useDatabase(): void
    {
        $database = $this->createDatabase();
        $database->connect();
        $database->query();
    }
}

final class FactoryMethodPostgresCreator extends FactoryMethodCreator
{
    protected function createDatabase(): FactoryMethodDatabase { return new FactoryMethodPostgres(); }
}

final class FactoryMethodMySqlCreator extends FactoryMethodCreator
{
    protected function createDatabase(): FactoryMethodDatabase { return new FactoryMethodMySql(); }
}

(new FactoryMethodPostgresCreator())->useDatabase();
(new FactoryMethodMySqlCreator())->useDatabase();
